pkg/collector/py/datadog_agent.c

```diff
@@ -101,12 +110,115 @@ static PyObject *get_subprocess_output(PyObject *self, PyObject *args) {
     return py_result;
 }
 
+static PyObject *add_external_tags(PyObject *self, PyObject *args) {
+    PyObject *input_list = NULL;
+    PyGILState_STATE gstate = PyGILState_Ensure();
+
+    // function expects only one positional arg containing a list
+    if (!PyArg_ParseTuple(args, "O", &input_list)) {
+        PyGILState_Release(gstate);
+        return NULL;
+    }
+
+    // if not a list, set an error
+    if (!PyList_Check(input_list)) {
+        PyErr_SetString(PyExc_TypeError, "function arg must be a list");
+        PyGILState_Release(gstate);
+        return NULL;
+    }
+
+    // if the list is empty do nothing
+    int input_len = PyList_Size(input_list);
+    if (input_len == 0) {
+        PyGILState_Release(gstate);
+        Py_RETURN_NONE;
+    }
+
+    int i;
+    for (i=0; i<input_len; i++) {
+        PyObject *tuple = PyList_GetItem(input_list, i);
+
+        // list must contain only tuples in form ('hostname', {'source_type': ['tag1', 'tag2']},)
+        if (!PyTuple_Check(tuple)) {
+            PyErr_SetString(PyExc_TypeError, "external host tags list must contain only tuples");
+            PyGILState_Release(gstate);
+            return NULL;
+        }
+
+        // first elem is the hostname
+        char *hostname = PyString_AsString(PyTuple_GetItem(tuple, 0));
```

Comment:
I think we should use 
```
const char * hostname  = PyString_AsString(PyTuple_GetItem(tuple, 0));
``` 
since `PyString_AsString()` points to the internal buffer and should never be modified.